### PR TITLE
Fix usage array function for CuPy modules

### PR DESCRIPTION
**Pull request by pentschev**
This PR solves https://github.com/cupy/cupy/issues/2011

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1242,12 +1242,14 @@ cdef class ndarray:
             return NotImplemented
 
     def __array_function__(self, func, types, args, kwargs):
-        if not hasattr(cupy, func.__name__):
+        module_list = '.'.join(func.__module__.split('.')[1:])
+        cupy_module = eval('cupy.' + module_list if module_list else 'cupy')
```
Comment (mrocklin) [inline cupy/core/core.pyx]: The `eval` function is often avoided, sometimes due to security (though I'm not sure that's relevant here) but more often just to be explicit.

I wonder if the following would work instead
```python
import cupy as module
for submodule in func.__module__.split('.')[1:]:
    module = getattr(module, submodule)
```

We should probably also catch any `ImportError` or `AttributeError`s and return `NotImplemented` if the module doesn't exist in CuPy.

(this is only one way to solve this problem.  I wouldn't be surprised if the cupy maintainers have other suggestions.)

Comment (niboshi) [reply cupy/core/core.pyx]: Thank you for PR & suggestion!
@mrocklin's suggestion looks good to me.
 Yes, we should catch `AttributeError`, but I guess not `ImportError` (`cupy` is already imported in the global scope).

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -1242,12 +1242,18 @@ cdef class ndarray:
             return NotImplemented
 
     def __array_function__(self, func, types, args, kwargs):
-        if not hasattr(cupy, func.__name__):
+        import cupy as module
```
Comment (niboshi) [inline cupy/core/core.pyx]: I think simply `module = cupy` shoud work.

Comment (pentschev) [reply cupy/core/core.pyx]: It does, thanks for spotting it. Pushing a commit in a minute.

Comment (niboshi) [reply cupy/core/core.pyx]: Thanks!

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
As a quick report, there're two failures from Jenkins:
```
19:28:31 ____________________ TestArrayFunction.test_array_function _____________________
19:28:31 
19:28:31 self = <cupy_tests.core_tests.test_array_function.TestArrayFunction testMethod=test_array_function>
19:28:31 
19:28:31     def test_array_function(self):
19:28:31         a = numpy.random.randn(100, 100)
19:28:31         a_cpu = numpy.asarray(a)
19:28:31         a_gpu = cupy.asarray(a)
19:28:31     
19:28:31         # The numpy call for both CPU and GPU arrays is intentional to test the
19:28:31         # __array_function__ protocol
19:28:31         qr_cpu = numpy.linalg.qr(a_cpu)
19:28:31 >       qr_gpu = numpy.linalg.qr(a_gpu)
19:28:31 
19:28:31 a          = array([[-0.31192251,  0.54540922, -1.95350922, ..., -2.15705981,
19:28:31         -1.00669283, -0.53702112],
19:28:31        [-0.9179094...205, -0.19056921],
19:28:31        [ 0.08367027,  0.86406915,  1.26840093, ...,  0.36209913,
19:28:31          0.37135528,  0.2503814 ]])
19:28:31 a_cpu      = array([[-0.31192251,  0.54540922, -1.95350922, ..., -2.15705981,
19:28:31         -1.00669283, -0.53702112],
19:28:31        [-0.9179094...205, -0.19056921],
19:28:31        [ 0.08367027,  0.86406915,  1.26840093, ...,  0.36209913,
19:28:31          0.37135528,  0.2503814 ]])
19:28:31 a_gpu      = array([[-0.31192251,  0.54540922, -1.95350922, ..., -2.15705981,
19:28:31         -1.00669283, -0.53702112],
19:28:31        [-0.9179094...205, -0.19056921],
19:28:31        [ 0.08367027,  0.86406915,  1.26840093, ...,  0.36209913,
19:28:31          0.37135528,  0.2503814 ]])
19:28:31 qr_cpu     = (array([[-0.02894301, -0.05205252, -0.20064207, ...,  0.08465095,
19:28:31          0.20440943,  0.01568078],
19:28:31        [-0.085171... [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00, ...,
19:28:31          0.00000000e+00,  0.00000000e+00,  2.00867216e-03]]))
19:28:31 self       = <cupy_tests.core_tests.test_array_function.TestArrayFunction testMethod=test_array_function>
19:28:31 
19:28:31 tests/cupy_tests/core_tests/test_array_function.py:21: 
19:28:31 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
19:28:31 /opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/numpy/linalg/linalg.py:903: in qr
19:28:31     a, wrap = _makearray(a)
19:28:31 /opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/numpy/linalg/linalg.py:117: in _makearray
19:28:31     new = asarray(a)
19:28:31 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
19:28:31 
19:28:31 a = array([[-0.31192251,  0.54540922, -1.95350922, ..., -2.15705981,
19:28:31         -1.00669283, -0.53702112],
19:28:31        [-0.9179094...205, -0.19056921],
19:28:31        [ 0.08367027,  0.86406915,  1.26840093, ...,  0.36209913,
19:28:31          0.37135528,  0.2503814 ]])
19:28:31 dtype = None, order = None
19:28:31 
19:28:31     @set_module('numpy')
19:28:31     def asarray(a, dtype=None, order=None):
19:28:31         """Convert the input to an array.
19:28:31     
19:28:31         Parameters
19:28:31         ----------
19:28:31         a : array_like
19:28:31             Input data, in any form that can be converted to an array.  This
19:28:31             includes lists, lists of tuples, tuples, tuples of tuples, tuples
19:28:31             of lists and ndarrays.
19:28:31         dtype : data-type, optional
19:28:31             By default, the data-type is inferred from the input data.
19:28:31         order : {'C', 'F'}, optional
19:28:31             Whether to use row-major (C-style) or
19:28:31             column-major (Fortran-style) memory representation.
19:28:31             Defaults to 'C'.
19:28:31     
19:28:31         Returns
19:28:31         -------
19:28:31         out : ndarray
19:28:31             Array interpretation of `a`.  No copy is performed if the input
19:28:31             is already an ndarray with matching dtype and order.  If `a` is a
19:28:31             subclass of ndarray, a base class ndarray is returned.
19:28:31     
19:28:31         See Also
19:28:31         --------
19:28:31         asanyarray : Similar function which passes through subclasses.
19:28:31         ascontiguousarray : Convert input to a contiguous array.
19:28:31         asfarray : Convert input to a floating point ndarray.
19:28:31         asfortranarray : Convert input to an ndarray with column-major
19:28:31                          memory order.
19:28:31         asarray_chkfinite : Similar function which checks input for NaNs and Infs.
19:28:31         fromiter : Create an array from an iterator.
19:28:31         fromfunction : Construct an array by executing a function on grid
19:28:31                        positions.
19:28:31     
19:28:31         Examples
19:28:31         --------
19:28:31         Convert a list into an array:
19:28:31     
19:28:31         >>> a = [1, 2]
19:28:31         >>> np.asarray(a)
19:28:31         array([1, 2])
19:28:31     
19:28:31         Existing arrays are not copied:
19:28:31     
19:28:31         >>> a = np.array([1, 2])
19:28:31         >>> np.asarray(a) is a
19:28:31         True
19:28:31     
19:28:31         If `dtype` is set, array is copied only if dtype does not match:
19:28:31     
19:28:31         >>> a = np.array([1, 2], dtype=np.float32)
19:28:31         >>> np.asarray(a, dtype=np.float32) is a
19:28:31         True
19:28:31         >>> np.asarray(a, dtype=np.float64) is a
19:28:31         False
19:28:31     
19:28:31         Contrary to `asanyarray`, ndarray subclasses are not passed through:
19:28:31     
19:28:31         >>> issubclass(np.recarray, np.ndarray)
19:28:31         True
19:28:31         >>> a = np.array([(1.0, 2), (3.0, 4)], dtype='f4,i4').view(np.recarray)
19:28:31         >>> np.asarray(a) is a
19:28:31         False
19:28:31         >>> np.asanyarray(a) is a
19:28:31         True
19:28:31     
19:28:31         """
19:28:31 >       return array(a, dtype, copy=False, order=order)
19:28:31 E       ValueError: object __array__ method not producing an array
19:28:31 
19:28:31 a          = array([[-0.31192251,  0.54540922, -1.95350922, ..., -2.15705981,
19:28:31         -1.00669283, -0.53702112],
19:28:31        [-0.9179094...205, -0.19056921],
19:28:31        [ 0.08367027,  0.86406915,  1.26840093, ...,  0.36209913,
19:28:31          0.37135528,  0.2503814 ]])
19:28:31 dtype      = None
19:28:31 order      = None
19:28:31 
19:28:31 /opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/numpy/core/numeric.py:538: ValueError
19:28:31 _______________________ TestDLTensorMemory.test_deleter ________________________
19:28:31 
19:28:31 self = <cupy_tests.core_tests.test_dlpack.TestDLTensorMemory testMethod=test_deleter>
19:28:31 
19:28:31     def test_deleter(self):
19:28:31         pool = cupy.get_default_memory_pool()
19:28:31         pool.free_all_blocks()
19:28:31         array = cupy.empty(10)
19:28:31         tensor = array.toDlpack()
19:28:31 >       assert pool.n_free_blocks() == 0
19:28:31 E       AssertionError: assert 2 == 0
19:28:31 E        +  where 2 = <built-in method n_free_blocks of cupy.cuda.memory.MemoryPool object at 0x7f42bac2bd30>()
19:28:31 E        +    where <built-in method n_free_blocks of cupy.cuda.memory.MemoryPool object at 0x7f42bac2bd30> = <cupy.cuda.memory.MemoryPool object at 0x7f42bac2bd30>.n_free_blocks
19:28:31 
19:28:31 array      = array([1.14937202e-001, 5.26889490e-002, 8.07467735e-001, 6.38091053e-001,
19:28:31        7.82251249e-001, 1.83094166e-001, 0.00000000e+000, 0.00000000e+000,
19:28:31        9.24424698e+081, 1.83456408e+206])
19:28:31 pool       = <cupy.cuda.memory.MemoryPool object at 0x7f42bac2bd30>
19:28:31 self       = <cupy_tests.core_tests.test_dlpack.TestDLTensorMemory testMethod=test_deleter>
19:28:31 tensor     = <capsule object "dltensor" at 0x7f4257329d20>
19:28:31 
19:28:31 tests/cupy_tests/core_tests/test_dlpack.py:40: AssertionError
```

**Comment by pentschev** [comment]
> As a quick report, there're two failures from Jenkins:
> 
> ```
> 19:28:31 ____________________ TestArrayFunction.test_array_function _____________________
> 19:28:31 
> 19:28:31 self = <cupy_tests.core_tests.test_array_function.TestArrayFunction testMethod=test_array_function>
> 19:28:31 
> 19:28:31     def test_array_function(self):
> 19:28:31         a = numpy.random.randn(100, 100)
> 19:28:31         a_cpu = numpy.asarray(a)
> 19:28:31         a_gpu = cupy.asarray(a)
> 19:28:31     
> 19:28:31         # The numpy call for both CPU and GPU arrays is intentional to test the
> 19:28:31         # __array_function__ protocol
> 19:28:31         qr_cpu = numpy.linalg.qr(a_cpu)
> 19:28:31 >       qr_gpu = numpy.linalg.qr(a_gpu)
> 19:28:31 
> 19:28:31 a          = array([[-0.31192251,  0.54540922, -1.95350922, ..., -2.15705981,
> 19:28:31         -1.00669283, -0.53702112],
> 19:28:31        [-0.9179094...205, -0.19056921],
> 19:28:31        [ 0.08367027,  0.86406915,  1.26840093, ...,  0.36209913,
> 19:28:31          0.37135528,  0.2503814 ]])
> 19:28:31 a_cpu      = array([[-0.31192251,  0.54540922, -1.95350922, ..., -2.15705981,
> 19:28:31         -1.00669283, -0.53702112],
> 19:28:31        [-0.9179094...205, -0.19056921],
> 19:28:31        [ 0.08367027,  0.86406915,  1.26840093, ...,  0.36209913,
> 19:28:31          0.37135528,  0.2503814 ]])
> 19:28:31 a_gpu      = array([[-0.31192251,  0.54540922, -1.95350922, ..., -2.15705981,
> 19:28:31         -1.00669283, -0.53702112],
> 19:28:31        [-0.9179094...205, -0.19056921],
> 19:28:31        [ 0.08367027,  0.86406915,  1.26840093, ...,  0.36209913,
> 19:28:31          0.37135528,  0.2503814 ]])
> 19:28:31 qr_cpu     = (array([[-0.02894301, -0.05205252, -0.20064207, ...,  0.08465095,
> 19:28:31          0.20440943,  0.01568078],
> 19:28:31        [-0.085171... [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00, ...,
> 19:28:31          0.00000000e+00,  0.00000000e+00,  2.00867216e-03]]))
> 19:28:31 self       = <cupy_tests.core_tests.test_array_function.TestArrayFunction testMethod=test_array_function>
> 19:28:31 
> 19:28:31 tests/cupy_tests/core_tests/test_array_function.py:21: 
> 19:28:31 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 19:28:31 /opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/numpy/linalg/linalg.py:903: in qr
> 19:28:31     a, wrap = _makearray(a)
> 19:28:31 /opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/numpy/linalg/linalg.py:117: in _makearray
> 19:28:31     new = asarray(a)
> 19:28:31 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 19:28:31 
> 19:28:31 a = array([[-0.31192251,  0.54540922, -1.95350922, ..., -2.15705981,
> 19:28:31         -1.00669283, -0.53702112],
> 19:28:31        [-0.9179094...205, -0.19056921],
> 19:28:31        [ 0.08367027,  0.86406915,  1.26840093, ...,  0.36209913,
> 19:28:31          0.37135528,  0.2503814 ]])
> 19:28:31 dtype = None, order = None
> 19:28:31 
> 19:28:31     @set_module('numpy')
> 19:28:31     def asarray(a, dtype=None, order=None):
> 19:28:31         """Convert the input to an array.
> 19:28:31     
> 19:28:31         Parameters
> 19:28:31         ----------
> 19:28:31         a : array_like
> 19:28:31             Input data, in any form that can be converted to an array.  This
> 19:28:31             includes lists, lists of tuples, tuples, tuples of tuples, tuples
> 19:28:31             of lists and ndarrays.
> 19:28:31         dtype : data-type, optional
> 19:28:31             By default, the data-type is inferred from the input data.
> 19:28:31         order : {'C', 'F'}, optional
> 19:28:31             Whether to use row-major (C-style) or
> 19:28:31             column-major (Fortran-style) memory representation.
> 19:28:31             Defaults to 'C'.
> 19:28:31     
> 19:28:31         Returns
> 19:28:31         -------
> 19:28:31         out : ndarray
> 19:28:31             Array interpretation of `a`.  No copy is performed if the input
> 19:28:31             is already an ndarray with matching dtype and order.  If `a` is a
> 19:28:31             subclass of ndarray, a base class ndarray is returned.
> 19:28:31     
> 19:28:31         See Also
> 19:28:31         --------
> 19:28:31         asanyarray : Similar function which passes through subclasses.
> 19:28:31         ascontiguousarray : Convert input to a contiguous array.
> 19:28:31         asfarray : Convert input to a floating point ndarray.
> 19:28:31         asfortranarray : Convert input to an ndarray with column-major
> 19:28:31                          memory order.
> 19:28:31         asarray_chkfinite : Similar function which checks input for NaNs and Infs.
> 19:28:31         fromiter : Create an array from an iterator.
> 19:28:31         fromfunction : Construct an array by executing a function on grid
> 19:28:31                        positions.
> 19:28:31     
> 19:28:31         Examples
> 19:28:31         --------
> 19:28:31         Convert a list into an array:
> 19:28:31     
> 19:28:31         >>> a = [1, 2]
> 19:28:31         >>> np.asarray(a)
> 19:28:31         array([1, 2])
> 19:28:31     
> 19:28:31         Existing arrays are not copied:
> 19:28:31     
> 19:28:31         >>> a = np.array([1, 2])
> 19:28:31         >>> np.asarray(a) is a
> 19:28:31         True
> 19:28:31     
> 19:28:31         If `dtype` is set, array is copied only if dtype does not match:
> 19:28:31     
> 19:28:31         >>> a = np.array([1, 2], dtype=np.float32)
> 19:28:31         >>> np.asarray(a, dtype=np.float32) is a
> 19:28:31         True
> 19:28:31         >>> np.asarray(a, dtype=np.float64) is a
> 19:28:31         False
> 19:28:31     
> 19:28:31         Contrary to `asanyarray`, ndarray subclasses are not passed through:
> 19:28:31     
> 19:28:31         >>> issubclass(np.recarray, np.ndarray)
> 19:28:31         True
> 19:28:31         >>> a = np.array([(1.0, 2), (3.0, 4)], dtype='f4,i4').view(np.recarray)
> 19:28:31         >>> np.asarray(a) is a
> 19:28:31         False
> 19:28:31         >>> np.asanyarray(a) is a
> 19:28:31         True
> 19:28:31     
> 19:28:31         """
> 19:28:31 >       return array(a, dtype, copy=False, order=order)
> 19:28:31 E       ValueError: object __array__ method not producing an array
> 19:28:31 
> 19:28:31 a          = array([[-0.31192251,  0.54540922, -1.95350922, ..., -2.15705981,
> 19:28:31         -1.00669283, -0.53702112],
> 19:28:31        [-0.9179094...205, -0.19056921],
> 19:28:31        [ 0.08367027,  0.86406915,  1.26840093, ...,  0.36209913,
> 19:28:31          0.37135528,  0.2503814 ]])
> 19:28:31 dtype      = None
> 19:28:31 order      = None
> 19:28:31 
> 19:28:31 /opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/numpy/core/numeric.py:538: ValueError
> 19:28:31 _______________________ TestDLTensorMemory.test_deleter ________________________
> 19:28:31 
> 19:28:31 self = <cupy_tests.core_tests.test_dlpack.TestDLTensorMemory testMethod=test_deleter>
> 19:28:31 
> 19:28:31     def test_deleter(self):
> 19:28:31         pool = cupy.get_default_memory_pool()
> 19:28:31         pool.free_all_blocks()
> 19:28:31         array = cupy.empty(10)
> 19:28:31         tensor = array.toDlpack()
> 19:28:31 >       assert pool.n_free_blocks() == 0
> 19:28:31 E       AssertionError: assert 2 == 0
> 19:28:31 E        +  where 2 = <built-in method n_free_blocks of cupy.cuda.memory.MemoryPool object at 0x7f42bac2bd30>()
> 19:28:31 E        +    where <built-in method n_free_blocks of cupy.cuda.memory.MemoryPool object at 0x7f42bac2bd30> = <cupy.cuda.memory.MemoryPool object at 0x7f42bac2bd30>.n_free_blocks
> 19:28:31 
> 19:28:31 array      = array([1.14937202e-001, 5.26889490e-002, 8.07467735e-001, 6.38091053e-001,
> 19:28:31        7.82251249e-001, 1.83094166e-001, 0.00000000e+000, 0.00000000e+000,
> 19:28:31        9.24424698e+081, 1.83456408e+206])
> 19:28:31 pool       = <cupy.cuda.memory.MemoryPool object at 0x7f42bac2bd30>
> 19:28:31 self       = <cupy_tests.core_tests.test_dlpack.TestDLTensorMemory testMethod=test_deleter>
> 19:28:31 tensor     = <capsule object "dltensor" at 0x7f4257329d20>
> 19:28:31 
> 19:28:31 tests/cupy_tests/core_tests/test_dlpack.py:40: AssertionError
> ```

Thanks for posting this, @niboshi. I forgot to mention before, but this PR is also related to https://github.com/cupy/cupy/pull/1650, and Jenkins will need to test with NumPy >= 1.16 and export `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1` before executing, this should fix the first error. Alternatively, I would suggest to add a `pytest.importorksip()` to run the test only if NumPy >= 1.17, it's still unreleased but `__array_function__` is enabled by default.

The second error, from `TestDLTensorMemory`, seems unrelated to me and I can't reproduce it here. Is it possible that this is hardware related or a non-deterministic bug?

**Comment by pentschev** [comment]
I fixed the conflict and took the liberty to skip the test if NumPy < 1.17. Please, let me know if that's acceptable.

**Comment by mrocklin** [comment]
Checking in.  Is there other work that remains to be done here?

**Comment by niboshi** [comment]
I'm sorry, we have been considering how we should test this feature.
Current conclusion is not to test it for now.
Could you just remove the test?

We will restore the test once NumPy 1.17 will be released.

**(For record)** Possibilities included:

1. Add a new CI matrix with `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1`.
2. Changing NumPy's internal flag from within the test.
3. Use a child process to run the test (with `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1`).


1\. is costly in terms of effort.
2\. just doesn't work; NumPy uses the flag on import.
3\. Child process would have to start in spawn mode, but that mode has to be configured globally; not appropriate as something to be done in a test.

**Comment by niboshi** [comment]
Ah, sorry, current test has `@testing.with_requires('numpy>=1.17.0')` so I think it's OK.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins failure is irrelevant to this PR.
`TestDLTensorMemory` failure is gone for some reason. I assume it was also irrelevant.

LGTM, thank you so much!

**Comment by mrocklin** [comment]
Thank you for the review @niboshi !

I'm looking forward to seeing better CuPy and Dask integration in the future.

Thanks @pentschev for finding and fixing this.

**Comment by pentschev** [comment]
Thanks @niboshi for taking the time to review. Thanks also @mrocklin for the help with reaching the solution.

**Pull request merged**